src/main/java/com/nimbusds/jose/crypto/ECDHDecrypterMulti.java

```diff
@@ -0,0 +1,174 @@
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A256CBC_HS512}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A128GCM}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A192GCM}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A256GCM}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A128CBC_HS256_DEPRECATED}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#A256CBC_HS512_DEPRECATED}
+ *     <li>{@link com.nimbusds.jose.EncryptionMethod#XC20P}
+ * </ul>
+ *
+ * @author Alexander Martynov
+ * @version 2021-08-19
+ */
+@ThreadSafe
+public class ECDHDecrypterMulti extends ECDHCryptoProvider implements JWEDecrypterMulti, CriticalHeaderParamsAware {
+
+
+    /**
+     * The supported EC JWK curves by the ECDH crypto provider class.
+     */
+    public static final Set<Curve> SUPPORTED_ELLIPTIC_CURVES;
+
+
+    static {
+        Set<Curve> curves = new LinkedHashSet<>();
+        curves.add(Curve.P_256);
+        curves.add(Curve.P_384);
+        curves.add(Curve.P_521);
+        SUPPORTED_ELLIPTIC_CURVES = Collections.unmodifiableSet(curves);
+    }
+
+    /**
+     * The critical header policy.
+     */
+    private final CriticalHeaderParamsDeferral critPolicy = new CriticalHeaderParamsDeferral();
+
+    private final ECKey[] recipients;
+
+    public ECDHDecrypterMulti(final ECKey[] recipients)
+            throws JOSEException {
+
+        this(recipients, null);
+    }
+
+    public ECDHDecrypterMulti(final ECKey[] recipients, final Set<String> defCritHeaders)
+        throws JOSEException {
+
+        super(recipients[0].getCurve());
+
+        this.recipients = recipients;
+        critPolicy.setDeferredCriticalHeaderParams(defCritHeaders);
+    }
+
+    @Override
+    public Set<Curve> supportedEllipticCurves() {
+
+        return SUPPORTED_ELLIPTIC_CURVES;
+    }
+
+
+    @Override
+    public Set<String> getProcessedCriticalHeaderParams() {
+
+        return critPolicy.getProcessedCriticalHeaderParams();
+    }
+
+
+    @Override
+    public Set<String> getDeferredCriticalHeaderParams() {
+
+        return critPolicy.getProcessedCriticalHeaderParams();
+    }
+
+
+    @Override
+    public byte[] decrypt(final JWEHeader header,
+                          final List<Recipient> recipients,
+                          final Base64URL iv,
+                          final Base64URL cipherText,
+                          final Base64URL authTag)
+            throws JOSEException {
+
+        critPolicy.ensureHeaderPasses(header);
+
+        // Get ephemeral EC key
+        ECKey ephemeralKey = (ECKey) header.getEphemeralPublicKey();
+
+        if (ephemeralKey == null) {
+            throw new JOSEException("Missing ephemeral public EC key \"epk\" JWE header parameter");
+        }
+
+        Map<String, SecretKey> sharedKeys = new HashMap<>();
+
+        for (ECKey recipient : this.recipients) {
+            if (!ECChecks.isPointOnCurve(ephemeralKey.toECPublicKey(), recipient.toECPrivateKey())) {
```

Comment:
Why don't move that check into `ECDH.deriveSharedSecret`?

Reply:
I don't want to change basic class behavior